src/backend/loadPlan.js

```diff
@@ -0,0 +1,50 @@
+//@flow
+
+import {Graph} from "../core/graph";
+import {type Project} from "../core/project";
+import {type TimelineCredParameters} from "../analysis/timeline/params";
+import {TimelineCred} from "../analysis/timeline/timelineCred";
+
+/**
+ * This file introduces a "plan" concept.
+ *
+ * A plan allows us to separate the large amount of dependency injection that's
+ * necessary to load a SourceCred instance. It makes a hard distinction between
+ * which arguments are environment-bound (like the TaskReporter, CacheProvider,
+ * and GithubToken) and which are parameters for the load (like Project).
+ *
+ * A <Something>Plan type only includes the later *parameters for loading*,
+ * because this Plan is typically the output of a factory function which curried
+ * the environment-bound arguments for us.
+ */
+
+export type MirrorPlan = (project: Project) => Promise<void>;
```

Comment:
It looks like this is supposed to silently have some side effect that is
depended on implicitly by some set of later processes. This kind of
spooky action at a distance tends to make systems harder to understand.

Instead, why doesn’t the `MirrorPlan` evaluate to a `Promise<Mirror>`?
Encode the information into the data; don’t suffer from [Boolean
blindness][bb].

[bb]: https://existentialtype.wordpress.com/2011/03/15/boolean-blindness/

Reply:
+1, I would find this a lot cleaner / easier to reason about.

Reply:
The main goal here is:
Both the concrete environment that was supplied *and* how the environment should be effected are abstracted at this level.

You're right, that makes the "Mirroring" operation rather abstract.
Because it *is* intended to side-effect, as the `Promise<void>` would typically communicate. But we're intentionally not aware of the environment, nor how it should be affected. The `Promise<void>` will simply tell us "it's done".

However, that's not the case for _concrete_ implementations. A concrete implementation also reveals what the environment is (the function wasn't partially applied yet), and how it should be affected.

When a higher level function constructs a `LoadPlan` instance. It references concrete implementations. So this function does is aware whether it would make sense to create dependencies through a shared environment here.

This is why `Promise<Mirror>` would be the environment leaking out, and a breach of this concept. The side-effects should remain in the environment. And a `LoadPlan` constructor is responsible for it's relationship.

Reply:
I have read your comment a few times but am having trouble following.
Specifically, I don’t understand the bits about “higher level”
functions, “affecting environments”, and “concrete implementations”.

Conceptually, I don’t see why the Mirror interface is “intended to
side-effect” any more than the others. Loading a mirror _may_ require
fetching from the network, or may fetch from cache, or may fetch from a
test database. It doesn’t matter. What matters are which dependencies we
need (here: a `Project`) and what outputs we provide (here: a `Mirror`).
The whole point of dependency injection is establish the coupling at
this semantic layer rather than the implementation layer.

If other parameters are needed—for instance, a GitHub token—shouldn’t
they also be in the dependency set rather than being hard-coded to
derive from environment variables?